### PR TITLE
feat(utils): add context to retry helper

### PR DESCRIPTION
**Pull request by ctrombley**

### CODE REVIEW

**Review by Julien4218 — APPROVED** [review]
@ctrombley this is great, exactly what I was hoping for.
One slight adjustment perhaps, I wonder if we could make the signature of ExecWithRetries( ) simpler by just returning the RetryContext without the last error.
It seems if the caller cares about error, the caller should fetch it directly in the returned RetryContext?
I could eventually see a Canceled bool in the RetryContext to capture that this was a canceled operation, to avoid having the consumer look at the last error to determine that it was. I also don't know we need to append an error when the operation is canceled if we capture the bool directly, just some thoughts.

File: internal/utils/retry.go [hunk 1]
```diff
@@ -37,13 +55,15 @@ func (r *Retry) ExecWithRetries(ctx context.Context) error {
 
 			select {
 			case <-ctx.Done():
-				return context.Canceled
+				retryCtx.Errors = append(retryCtx.Errors, context.Canceled)
```
Comment (Julien4218) [inline internal/utils/retry.go:58]: @ctrombley should we update retryCtx.Canceled = true when context.Canceled?

File: internal/diagnose/config_validator.go [hunk 2]
```diff
@@ -108,7 +110,13 @@ func (c *ConfigValidator) validateKeys(ctx context.Context) error {
 	}
 
 	r := utils.NewRetry(c.PostMaxRetries, c.PostRetryDelaySec*1000, validateKeyFunc)
-	return r.ExecWithRetries(ctx)
+	retryCtx := r.ExecWithRetries(ctx)
+
+	if !retryCtx.Success {
+		return retryCtx.Errors[len(retryCtx.Errors)-1]
```
Comment (Julien4218) [inline internal/diagnose/config_validator.go:116]: @ctrombley maybe use the new method MostRecentError( ) here too

File: internal/utils/retry.go (hunk 1 above)
Comment (ctrombley) [reply internal/utils/retry.go:58]: 👍 I missed this, thanks.

**Pull request merged**